File: src/main/java/io/curiousoft/ijudi/ordermanagement/model/GeoDistance.java
package io.curiousoft.ijudi.ordermanagement.model;

public abstract class GeoDistance implements Comparable<GeoDistance> {

    public static final double EARTH_RADIUS_IN_KM = 6378.137;
    public static final double TO_RADIANS = Math.PI / 180;
    private double distance;
    private double direction;


    /**
     * Uses Harvesine formula to calculate the distance between two points in a sphere
     *
     * @return The distance between points
     */
    public static double getDistanceInKiloMetersBetweenTwoGeoPoints(GeoPoint a, GeoPoint b) {

        double latitude = a.getLatitude() * TO_RADIANS;
        double longitude = a.getLongitude() * TO_RADIANS;
        double latitude1 = b.getLatitude() * TO_RADIANS;
        double longitude1 = b.getLongitude() * TO_RADIANS;

        double results = Math.pow(Math.sin((latitude1 - latitude) / 2), 2) + (Math.cos(latitude) * Math.cos(latitude1) * Math
                .pow(Math.sin((longitude1 - longitude) / 2), 2));
        double dist = 2 * Math.atan2(Math.sqrt(results), Math.sqrt(1 - results));
        return dist * EARTH_RADIUS_IN_KM;
    }

    public static double getAngleInDegreesBetweenTwoGeoPoints(GeoPoint a, GeoPoint b) {
        double angle = Math.toDegrees(
                Math.atan2(b.getLatitude() - a.getLatitude(), b.getLongitude() - a.getLongitude()));
        return angle >= 0 ? angle : 360 + angle;
    }

    public double getDistance() {
        return distance;
    }

    /**
     * @return an angle between A and B in degrees.
     */
    public double getDirection() {
        return direction;
    }

    public void setDistance(double distance) {
        this.distance = distance;
    }

    @Override
    public int compareTo(GeoDistance o) {
        return distance > o.getDistance() ? 1 : distance == o.getDistance() ? 0 : -1;
    }
}

File: izinga-commons/src/main/kotlin/io/curiousoft/izinga/commons/model/GeoDistance.kt
package io.curiousoft.izinga.commons.model

abstract class GeoDistance : Comparable<GeoDistance> {
    var distance = 0.0

    /**
     * @return an angle between A and B in degrees.
     */
    val direction = 0.0
    override fun compareTo(o: GeoDistance): Int {
        return if (distance > o.distance) 1 else if (distance == o.distance) 0 else -1
    }

    companion object {
        const val EARTH_RADIUS_IN_KM = 6378.137
        const val TO_RADIANS = Math.PI / 180

        /**
         * Uses Harvesine formula to calculate the distance between two points in a sphere
         *
         * @return The distance between points
         */
        fun getDistanceInKiloMetersBetweenTwoGeoPoints(a: GeoPoint, b: GeoPoint): Double {
            val latitude = a.latitude * TO_RADIANS
            val longitude = a.longitude * TO_RADIANS
            val latitude1 = b.latitude * TO_RADIANS
            val longitude1 = b.longitude * TO_RADIANS
            val results =
                Math.pow(Math.sin((latitude1 - latitude) / 2), 2.0) + Math.cos(latitude) * Math.cos(latitude1) * Math
                    .pow(Math.sin((longitude1 - longitude) / 2), 2.0)
            val dist = 2 * Math.atan2(Math.sqrt(results), Math.sqrt(1 - results))
            return dist * EARTH_RADIUS_IN_KM
        }

        fun getAngleInDegreesBetweenTwoGeoPoints(a: GeoPoint, b: GeoPoint): Double {
            val angle = Math.toDegrees(
                Math.atan2(b.latitude - a.latitude, b.longitude - a.longitude)
            )
            return if (angle >= 0) angle else 360 + angle
        }
    }
}